### PR TITLE
add CTrL integration

### PR DESCRIPTION
**Pull request by TomVeniat**
Draft for #377

### CODE REVIEW

**Comment by vlomonaco** [comment]
Hi @TomVeniat let us know when we can start to review this and please add some comments in the PR description so that we can understand the logic behind it! thanks! :)

**Comment by TomVeniat** [comment]
Hi @vlomonaco,
I think we're almost done, where could I add a demo of how to use it?
I think a notebook would be great but it seems notebooks are only used for the "from zero to hero" series and that all other examples are scripts, is that correct?

**Comment by vlomonaco** [comment]
Hi Tom, you can add a comprehensive example on its usage in the `examples` directory. Also please add unit tests, doc strings and try to add non only the "classic" CtRL Benchmark but also expose its "internal" datasets, so they can be re-used with the avalanche benchmark generators, thanks!

**Comment by AntonioCarta** [comment]
Maybe a better place for a notebook would be outside of avalanche's documentation, in our [colab repository](https://github.com/ContinualAI/colab). We have a collection of notebooks and continual learning examples there.

What do you think?

**Comment by TomVeniat** [comment]
I think that only exposing the "classic" streams is a good starting point, with a simple script in the `examples` directory (similar to https://github.com/ContinualAI/avalanche/blob/master/examples/all_mnist.py), just to show how to get started.  
Once this is done I'll work on exposing the internals and making a notebook to demonstrate how to create new streams, as it will require a bit more work on the CTrL side. Does this sound good to you?

About the testing, is there a way to add some information about each experience in a `dataset_benchmark`? I'd like to be able to verify that the same category appears in different tasks, so adding the class names in each experience could be useful.
Another use case would be task descriptors, is it possible to use them in the current version of Avalanche?
Thanks for your help !

**Comment by vlomonaco** [comment]
> I think that only exposing the "classic" streams is a good starting point, with a simple script in the examples directory (similar to https://github.com/ContinualAI/avalanche/blob/master/examples/all_mnist.py), just to show how to get started.
Once this is done I'll work on exposing the internals and making a notebook to demonstrate how to create new streams, as it will require a bit more work on the CTrL side. Does this sound good to you?

Yes, it makes sense thanks! :)

> About the testing, is there a way to add some information about each experience in a dataset_benchmark? I'd like to be able to verify that the same category appears in different tasks, so adding the class names in each experience could be useful.
Another use case would be task descriptors, is it possible to use them in the current version of Avalanche?

You can check the "[classes_in_experience](https://avalanche-api.continualai.org/autoapi/avalanche/benchmarks/scenarios/generic_cl_scenario/?highlight=scenario#avalanche.benchmarks.scenarios.generic_cl_scenario.GenericCLScenario)" which is populated automatically. As for tasks descriptors unfortunately I'm not sure we currently support that as `task_labels` and are only integers across avalanche at the moment (can you confirm this @lrzpellegrini?).

**Comment by lrzpellegrini** [comment]
I'm afraid that `AvalancheDataset` currently supports only int `task_labels`. Having a task descriptor field in `AvalancheDataset` (as a separate `task_descriptor` field?) is a good idea we already thinked of (we also have a pending feature request #371), but that feature remained low on the priority list... It should't be too difficult to implement this feature in its plain version where we have one descriptor per pattern (just like the `targets` and `task_labels` fields) and the descriptors are of arbitrary types (Tensors, ndarrays, scalars).

@TomVeniat you need `Tensor` descriptors, am I correct?

**Comment by vlomonaco** [comment]
@TomVeniat any update on this?

**Comment by TomVeniat** [comment]
Hi @vlomonaco,
Yes, the integration of existing streams, with doc and tests is ready. 
I also have a demo script that trains a simple CNN on a whole stream using the `Naive` strategy. I started working on the task-level cross-validation but got caught up with other things the last few weeks, sorry about that!
The PR will be ready for review as soon as this part is done, I think it is important to provide an example of the correct way to evaluate any model on CTrL.
You can expect an update this week, would that work for you ?

**Comment by vlomonaco** [comment]
Yes of course, thanks @TomVeniat! :) Let us know if you need any help!

File: avalanche/benchmarks/classic/ctrl.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            dataset = AvalancheTensorDataset(samples, labels.squeeze(1),
+                                             task_labels=task_labels,
+                                             transform=trans)
+            exp.append(dataset)
+
+    return dataset_benchmark(
+        train_datasets=exps[0],
+        test_datasets=exps[2],
+        other_streams_datasets=dict(valid=exps[1]),
+    )
```
Comment (lrzpellegrini) [inline avalanche/benchmarks/classic/ctrl.py:36]:
```suggestion
    )


__all__ = [
    'CTrL'
]

```

File: examples/simple_ctrl.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--mnist_type', type=str, default='split',
+                        choices=['rotated', 'permuted', 'split'],
+                        help='Choose between MNIST variations: '
+                             'rotated, permuted or split.')
+    parser.add_argument('--cuda', type=int, default=0,
+                        help='Select zero-indexed cuda device. -1 to use CPU.')
+    args = parser.parse_args()
+
+    main(args)
```
Comment (lrzpellegrini) [inline examples/simple_ctrl.py:115]: Hi Tom, can you clean up the example a little bit (remove unused imports, change the file description, ...)?

**Review by lrzpellegrini — CHANGES_REQUESTED** [review]
Hi Tom, apart from the comments I left in `ctrl.py` and `simple_ctrl.py`, can you add ctrl in the `__init__.py` file in `avalanche/benchmarks/classic` (following the alphabetical order of modules)?

**Comment by TomVeniat** [comment]
Thank you @lrzpellegrini for the review, I'll make the suggested changes.

The only thing missing for the full integration of CTrL is a way to handle longer streams. The current integration starts with loading all experiences as `AvalancheTensorDataset`s, and then creates the benchmark using `dataset_benchmark`.
The problem is that the long stream of CTrL doesn't necessarily fit in memory, preventing the usage of this method. 
Is there a way to lazily generate the Experiences in a benchmark? 
An alternative solution would be to write each experience generated with CTrL on the disk and only then create the benchmark with a list of paths. Is this possible in the current version? If it isn't, do you think that writing every sample to the disk and then use `path_benchmark` work?

I will also open a new issue for the task-level cross-validation, the solution I had isn't that straightforward to implement so Id' like your input on it to make sure it fits nicely with the framework.

**Comment by lrzpellegrini** [comment]
> Is there a way to lazily generate the Experiences in a benchmark?

Alas, in the current state this is not possible. However, this is a feature we already have in our backlog (see #600).

> An alternative solution would be to write each experience generated with CTrL on the disk and only then create the benchmark with a list of paths. Is this possible in the current version? If it isn't, do you think that writing every sample to the disk and then use path_benchmark work?

Yes, this is already doable. Have a look at: https://github.com/ContinualAI/avalanche/blob/68290d32f9710639d5fda6d17da4072a4458f702/avalanche/benchmarks/scenarios/generic_benchmark_creation.py#L270 (which is the alias for `paths_benchmark`). That helper accepts a list of files for each experience and it also supports custom streams (like the validation one). In that case, the best solution is to write any generated data to a temporary folder or some other kind of "working directory".

Avalanche doesn't currently support directly loading tensor(s) describing a whole experience directly from disk, but I think I can easily implement it if needed!

> I will also open a new issue for the task-level cross-validation, the solution I had isn't that straightforward to implement so Id' like your input on it to make sure it fits nicely with the framework.

Yes, a separate issue seems the best solution. For that feature we'll need some support from @AntonioCarta on the strategy-side.

**Comment by TomVeniat** [comment]
> Yes, this is already doable. Have a look at:
> 
> https://github.com/ContinualAI/avalanche/blob/68290d32f9710639d5fda6d17da4072a4458f702/avalanche/benchmarks/scenarios/generic_benchmark_creation.py#L270
>(which is the alias for paths_benchmark). That helper accepts a list of files for each experience and it also supports custom streams (like the validation one). In that case, the best solution is to write any generated data to a temporary folder or some other kind of "working directory".

I started the implementation following this strategy but It seems that there is no way to give the correct transform to each experience. The `train_transform` and `eval_transform` attributes of `path_benchmark` only accept a global transform for the whole benchmark and can't work with lists of transforms (one per experience) in the current version. Am I overlooking something?
I think the best alternative would be to create each `PathsDataset` manually with the correct transform and then create the benchmark using `dataset_benchmark` instead of `path_benchmark`.

**Comment by lrzpellegrini** [comment]
Yes, if a different transformation is needed for each experience, then `dataset_benchmark` is the best solution!

**Comment by lrzpellegrini** [comment]
Hi Tom, I recently implemented a minimal support for lazily generated streams, see #671.

That approach may make things easier on your side! There is also some memory optimization tuning that can be used to prevent memory usage from exploding by dropping references to previous experiences (I still have to document it decently). That PR still has to be merged, but it should be a matter of hours.

**Comment by TomVeniat** [comment]
Thanks for the update! I should have posted here but I went the write to disk + `dataset_benchmark` route for the long stream in [this commit](https://github.com/ContinualAI/avalanche/pull/561/commits/ec7727878a1a99002122cca9c5d3d9715f0af997).
I think this PR is ready for a review, the last thing to do will be to remove the EarlyStopping feature and replace it with the official one when #670 is merged.

**Comment by TomVeniat** [comment]
It seems like the memory consumption is still too high for the long stream, I will need to make some changes on the ctrl package side to make it more efficient (I think that I keep a reference to all generated tasks so far during the generation process).
In the meantime, I added a parameter controlling the length of the long stream for more flexibility and I use it in the tests to make them run faster and relieve the memory pressure.

**Comment by TomVeniat** [comment]
@lrzpellegrini Any idea of what happened with the python 3.9 tests setup?

**Comment by ashok-arjun** [comment]
@TomVeniat This seems to be the error:

```
Traceback (most recent call last):
  File "/__w/avalanche/avalanche/tests/test_ctrl.py", line 63, in test_determinism
    bench_1 = CTrL(stream, seed=1)
  File "/__w/avalanche/avalanche/avalanche/benchmarks/classic/ctrl.py", line 55, in CTrL
    stream = ctrl.get_stream(stream_name, seed)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/streams/__init__.py", line 144, in get_stream
    return init_component(default_rng(seed), **config)['task_gen']
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/streams/__init__.py", line 130, in init_component
    v = init_component(_rnd=_rnd, **v)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/streams/__init__.py", line 130, in init_component
    v = init_component(_rnd=_rnd, **v)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/streams/__init__.py", line 130, in init_component
    v = init_component(_rnd=_rnd, **v)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/streams/__init__.py", line 134, in init_component
    return comp_class(seed=_rnd.integers(0, 1e9), **kwargs)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/instances/image_dataset_tree.py", line 120, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/concepts/concept_tree.py", line 57, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/commons/tree.py", line 27, in __init__
    self.root_node = self.build_tree()
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/instances/image_dataset_tree.py", line 129, in build_tree
    self.trainset = ds_class(split='train', **common_params)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/instances/DTD.py", line 41, in __init__
    self._download_and_prepare(root, split, img_size)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/ctrl/instances/DTD.py", line 47, in _download_and_prepare
    download_and_extract_archive(self.url, download_root=root)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/torchvision/datasets/utils.py", line 417, in download_and_extract_archive
    extract_archive(archive, extract_root, remove_finished)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/torchvision/datasets/utils.py", line 383, in extract_archive
    suffix, archive_type, compression = _detect_file_type(from_path)
  File "/opt/conda/envs/avalanche-env/lib/python3.9/site-packages/torchvision/datasets/utils.py", line 305, in _detect_file_type
    raise RuntimeError(
RuntimeError: Archive type and compression detection only works for 1 or 2 suffixes. Got 4 instead.
```


This is from the `ctrl-benchmark`, specifically the `download_and_extract_archive` does not work with `torchvision` anymore. This prevents the download of the `dtd` dataset.

I got the same error when running on `python 3.9` on multiple machines.

I have a fix with which I am working locally, and to integrate that, you need to merge the [PR](https://github.com/facebookresearch/CTrLBenchmark/pull/4) I raised at [facebookresearch/CTrLBenchmark](https://github.com/facebookresearch/CTrLBenchmark/).

**Comment by vlomonaco** [comment]
Thanks @ashok-arjun for this! :)

**Comment by TomVeniat** [comment]
Thanks @ashok-arjun, your PR is live! :)
@vlomonaco, is there anything I missed to force the test pipeline to take into account the changes made in `environment.yml` ? The failures are due to `ctrl-benchmark` not being installed when the tests are run.

**Comment by ashok-arjun** [comment]
@lrzpellegrini Any idea why these tests aren't passing? I am not able to figure this out.

**Comment by AndreaCossu** [comment]
Hi @ashok-arjun, is your repo in sync with avalanche master branch? If not, can you please update it (there may be conflicts to be solved)?

**Comment by ashok-arjun** [comment]
@AndreaCossu Actually it is @TomVeniat 's branch that must be checked for conflicts, since he has made the PR :)

**Comment by ashok-arjun** [comment]
The error related to the unit test failures is `ModuleNotFoundError: No module named 'ctrl'`.

**Comment by TomVeniat** [comment]
Thanks @ashok-arjun  for looking into this issue !

> [...] Is there anything I missed to force the test pipeline to take into account the changes made in `environment.yml` ? The failures are due to `ctrl-benchmark` not being installed when the tests are run.

It seems the error remains the same, with the updates of `environment-dev.yml` and `environment.yml` not taken into account when running the tests. 
Is there anything I can do to help with that ?

**Comment by AndreaCossu** [comment]
I added the `ctrl-benchmark` to the docker test. Thanks @TomVeniat and @ashok-arjun !
If @lrzpellegrini is ok with the changes, we can merge this.

**Comment by lrzpellegrini** [comment]
Everything seems in order to me!

**Pull request merged**